src/core/src/services/formly.form.expression.spec.ts

```diff
@@ -1,15 +1,281 @@
 import { FormlyConfig } from './formly.config';
-import { Validators, FormGroup } from '@angular/forms';
+import { FormlyFieldConfig, FormlyFormOptions } from '../core';
+import { AbstractControl, FormArray, FormControl, FormGroup, Validators } from '@angular/forms';
 import { Component } from '@angular/core';
+import { evalStringExpression, evalExpressionValueSetter, FORMLY_VALIDATORS, getKeyPath, isFunction, isObject } from './../utils';
 
 import { FormlyFormExpression } from './formly.form.expression';
 
+function registerFormControls(form: FormGroup | FormArray, fields: FormlyFieldConfig[], model: any, options: FormlyFormOptions) {
```

Comment:
I'd prefer to use `FormlyFormBuilder` to build form instead

Reply:
One concern I had with using `FormlyFormBuilder` is that the public `buildForm` inherently calls FormlyFormExpression [`this.formlyFormExpression.checkFields(form, fields, model, options);`](https://github.com/formly-js/ngx-formly/blob/master/src/core/src/services/formly.form.builder.ts#L34).

So tests would look like this:

```
    builder.buildForm(form, fields, model, options);

    // expression.checkFields(form, fields, model, options); <-- inherently called but could be called explicitly called again

    expect(fields[1].hide).toBeTruthy();
    expect(fields[1].templateOptions.hidden).toBeTruthy();
```

As I'm typing this, what I could do, is initialize the FormBuilder and ignore it's internal instance of `formlyFormExpression` and have a separate instance just for the test.

Does this make sense? 😜

Reply:
But I agree I don't like having duplicate functions in the test itself.

Reply:
you can pass a mocked instance instead to avoid call checkFields inside FormlyFormBuilder.

Reply:
Yes, I'll have something this afternoon! 👍